llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -1434,6 +1436,33 @@ static bool checkAndReplaceMinMax(MinMaxIntrinsic *MinMax, ConstraintInfo &Info,
   return false;
 }
 
+static bool checkAndReplaceCmp(IntrinsicInst *II, ConstraintInfo &Info,
+                               SmallVectorImpl<Instruction *> &ToRemove) {
+  bool IsSigned = II->getIntrinsicID() == Intrinsic::scmp;
+  Value *LHS = II->getOperand(0);
+  Value *RHS = II->getOperand(1);
+  if (checkCondition(IsSigned ? ICmpInst::ICMP_SGT : ICmpInst::ICMP_UGT, LHS,
+                     RHS, II, Info)
+          .value_or(false)) {
+    II->replaceAllUsesWith(ConstantInt::get(II->getType(), 1));
+    ToRemove.push_back(II);
+    return true;
+  }
+  if (checkCondition(IsSigned ? ICmpInst::ICMP_SLT : ICmpInst::ICMP_ULT, LHS,
+                     RHS, II, Info)
+          .value_or(false)) {
+    II->replaceAllUsesWith(ConstantInt::getSigned(II->getType(), -1));
+    ToRemove.push_back(II);
+    return true;
+  }
+  if (checkCondition(ICmpInst::ICMP_EQ, LHS, RHS, II, Info).value_or(false)) {
```

Comment:
We can avoid this check since `!(x < y) && !(x > y)` implies `x == y`.

```suggestion
  auto CmpGt = checkCondition(IsSigned ? ICmpInst::ICMP_SGT : ICmpInst::ICMP_UGT, LHS,
                     RHS, II, Info);
  if (CmpGt.value_or(false)) {
    II->replaceAllUsesWith(ConstantInt::get(II->getType(), 1));
    ToRemove.push_back(II);
    return true;
  }
  auto CmpLt = checkCondition(IsSigned ? ICmpInst::ICMP_SLT : ICmpInst::ICMP_ULT, LHS,
                     RHS, II, Info);
  if (CmpLt.value_or(false)) {
    II->replaceAllUsesWith(ConstantInt::getSigned(II->getType(), -1));
    ToRemove.push_back(II);
    return true;
  }
  if (CmpLt == false && CmpGt == false) {
```

Reply:
It's true in principle, but implementing it this way wouldn't fold specifically `scmp` in situations where equality is implied.
E.g. this test case doesn't fold:
```
define i8 @test(i32 noundef %x, i32 noundef %y) {
  %cond = icmp eq i32 %x, %y
  call void @llvm.assume(i1 %cond)
  %r = call i8 @llvm.scmp.i8.i32(i32 %x, i32 %y)
  ret i8 %r
}
```
I'm not sure why this is the case, but it looks like when constraint elimination state is fed an assumption that `x == y` it adds the two implied inequalities to its unsigned table, but not to the signed one. That's also why the same test above will fold nicely if you replace `scmp` with `ucmp`. Unless I'm missing something and this implication isn't necessarily correct I'm happy to change my code as per your suggestion and fix the contraint elimination pass in as a whole in a separate PR.

Reply:
@fhahn Do you have a plan to support eq/ne with signed predicates?

```
define i1 @test1(i32 noundef %x, i32 noundef %y) {
  %cond = icmp eq i32 %x, %y
  call void @llvm.assume(i1 %cond)
  %r = icmp slt i32 %x, %y
  ret i1 %r
}

define i1 @test2(i32 noundef %x, i32 noundef %y) {
  %cond1 = icmp sle i32 %x, %y
  call void @llvm.assume(i1 %cond1)
  %cond2 = icmp sge i32 %x, %y
  call void @llvm.assume(i1 %cond2)
  %r = icmp eq i32 %x, %y
  ret i1 %r
}
```

Reply:
I think that, for the first example, it could suffice to check whether the equality holds and change the predicate to unsigned. Compile-time does not look good though: https://llvm-compile-time-tracker.com/compare.php?from=7911fb1a257b3a7014b44b4e7d04ee5c3b73a3e3&to=612eaf91c85d5e4c6e3fdddd46ebe6a41d19dfcb&stat=instructions:u. Perhaps there's a better way to achieve this?

Reply:
For the second test case the simplest solution would probably be checking `EQ`/`NE` predicates against both constraint system. I'm not sure about the performace impact though and don't reallly know how to check it.

Reply:
> I think that, for the first example, it could suffice to check whether the equality holds and change the predicate to unsigned. Compile-time does not look good though: https://llvm-compile-time-tracker.com/compare.php?from=7911fb1a257b3a7014b44b4e7d04ee5c3b73a3e3&to=612eaf91c85d5e4c6e3fdddd46ebe6a41d19dfcb&stat=instructions:u. Perhaps there's a better way to achieve this?

`doesHold(CmpInst::ICMP_EQ, Op0, Op1)` will build another constraint system to solve. It's no surprise that the optimization slows down when you do this.
It would be better to add eq constraint into the signed system in `ConstraintInfo::addFact`.

Sorry about the misleading emoji :)

Reply:
> It would be better to add eq constraint into the signed system in ConstraintInfo::addFact.

Oh, sure! Maybe adding sge and sle facts if eq predicate in `transferToOtherSystem` could work too.